### PR TITLE
Add OpenIddict tables to database

### PR DESCRIPTION
**Pull request by bergmania**
## Details
This PR adds database tables for OpenIddict using EF Core.

The tables are created doing install using a new notification (`DatabaseSchemaAndDataCreatedNotification`). This is required due to the fact SQLite migrations needs to run outside of a transaction.
Therefore I implemented a new notification handler that handles both the new `DatabaseSchemaAndDataCreatedNotification` and the existing `UnattendedInstallNotification`. Thereby creating the tables outside of the transaction.

For upgrades, I added a new Umbraco migration that executes a single migration from EF Core.
I had to abstract alot of things to make it work in the recommended way for[ EF Core to use a single context with multiple database providers](https://learn.microsoft.com/en-us/ef/core/managing-schemas/migrations/providers?tabs=dotnet-core-cli#using-one-context-type) .

The short version is migrations for each provider needs to be in its own assemblies.

The OpenIddict implementation is as small as I could, so this can be on it's own PR.

## Tests
 Ensure the OpenIddict tables (`umbracoOpenIddictApplications`, `umbracoOpenIddictAuthorizations`, `umbracoOpenIddictScopes` & `umbracoOpenIddictTokens`) and EFCore history (`__EFMigrationsHistory`) table is created on **_install_**,  _**unattended Installs**_ and _**upgrade**_ using both Sqlite and SqlServer.

### CODE REVIEW

**Comment by nikcio** [comment]
# ⚠️ Concerns ⚠️ 

Hey, @bergmania and @nikolajlauridsen I've been tinkering with adding the current database models from Umbraco to an EF Core DbContext and have in that process encountered the changes in this PR. Therefore, I would like to express some concerns I have about how this feature is implemented.

## 🔴 Migrations 🔴 

First up is the migrations here you've added a migration named `InitialCreate` which adds the needed models for OpenIddict without first making a migration with the existing models in the database. This means that if you were to want later to incorporate the other existing models to the `UmbracoDbContext` you will have migrations that have the following flow:

NPOCO migrations --> OpenIddict migration (InitialCreate) --> ... --> NPOCO migration --> Other EF Core migrations

A better flow in my opinion is to make a clean cut from NPOCO migrations to EF Core migrations because this makes us able to use the scaffolding tool to create the initial setup and saves a ton of time getting it to work. This flow would look like this.

NPOCO migrations --> InitialCreate (To update the DbSnapshot with the models from the NPOCO migrations) --> AddOpenIddict (The migration that is now named `InitalCreate`) --> Future EF Core migrations

This tactic avoids having developers use two migration frameworks at the same time which is sure to cause confusion and even further postpone a full migration to only EF Core migrations. This is because while having parallel flows of migration frameworks the NPOCO migrations will still have the same presence in Umbraco.

### A future with a full migration to EF Core

So how would you in the future completely migrate to EF Core without migrating the current NPOCO migrations to EF Core ones?
This would simply be to have the NPOCO migrations for x number of years/versions and then have a clean cut at some point where the NPOCO migrations can be deleted.

Group 1 (1)

Consider the example diagram above. Here the NPOCO migrations will be in the Umbraco CMS until `v15` and after that be deleted. This will mean that people upgrading would have to upgrade to any version between v12 and v15 before upgrading to future versions. This will though only be possible because of the overlap period of the NPOCO migrations and the EF Core migrations where there aren't added any NPOCO migrations (Marked with the brown color). This also means that if you in the overlapping period make new NPOCO migrations you need to extend the overlap period further to avoid having developers be forced to upgrade to a specific version before upgrading further. So TLDR making a clean switch where the least upgrades are affected is to make a grace period where the NPOCO migrations are still present but all new migrations are done by EF Core.

### Migration naming

In this PR you have created a migration that is called `InitalCreate` which doesn't clearly reflect what the migration does which is adding models for OpenIddict. A better name would be to use `AddOpenIddict`. Migration names are important when other developers have to figure out what happened when and when you sit in the future and try to figure out what happened when.

### Migration snapshot name

In the migration snapshot, the class is called `UmbracoOpenIddictDbContextModelSnapshot` but the DbContext is called `UmbracoDbContext` these two models should follow the same naming scheme to avoid confusion and clearly show what snapshot belongs to what DbContext.

## DbContext

The name `UmbracoDbContext` suggests that the DbContext should be used for the future of the Umbraco repositories that interact with the database and not only `OpenIddict` if that was what that was planned. (Observation based on the other changes described above).

### OnModelCreating

In the `OnModelCreating` method you have created an automatic prefix of `umbraco` to every table in the DbContext. This is in itself fine until you consider the current tables in the database prefixed with `cms`. Therefore, I wouldn't make such a constraint before all tables share the same prefix.

### Missing `DbSet`

In the implementation of this PR the tables added from OpenIddict isn't directly obvious from the DbContext which makes the tables almost magic-like (This is bad in this context). In my experience working with EF Core, you want to have as declarative code as possible to enhance the possibility to debug the code and for other developers to understand how the tables in the database are set up. This also makes it much easier in the future if you want to modify a table in some way because you can see where all the models come from and what they look like.
A solution to this problem is therefore to create models in the CMS that will reflect the models used by OpenIddict and with that models create the configuration needed.

## DbContext services

The `AddUmbracoEFCoreContext` method added to `UmbracoEFCoreServiceCollectionExtensions.cs` adds unpooled DbContexts. This is a problem because the default DbContext has some severe performance limitations compared to using a pooled DbContext factory therefore the default `DbContext` shouldn't be registered in the service collection because it encourages slower applications compared to the `IDbContextFactory<TContext>` you can inject everywhere because it's an singleton. Therefore the `AddDbContextFactory` also needs to be replaced with the `AddPooledDbContext` call.

See more here: https://learn.microsoft.com/en-us/ef/core/performance/advanced-performance-topics?tabs=with-di%2Cexpression-api-with-constant

### `AddUmbracoEFCoreContext` adjacent methods

Also, a small note the two `AddUmbracoEFCoreContext` methods should be adjacent.

## Provider setup

The calls to `UseSqlite` and `UseSqlServer` should include a strongly defined path to the migration assembly that they are using. This is again to improve traceability and avoid magic-like code.

## Umbraco.Cms.Api.Common project references

This project shouldn't directly reference the infrastructure layer like this. This makes for a complicated solution setup that is hard to understand and right now there's a transitive reference through the Umbraco.Web.Common project. See #13593

## Umbraco.Cms.Api.Delivery/DependencyInjection/UmbracoBuilderExtensions.cs

Here there's been added a direct call to use the EF Core DbContext when using the delivery API. This doesn't make any sense as the DbContext should always be available when the rest of the setup for this is done with the composer in Umbraco.Cms.Persistence.EFCCore. This is also needed if we are going to use this DbContext for anything else than OpenIddict like the name suggests.

**Comment by nikcio** [comment]
I've also created a PR showing how EF Core could be integrated so that it could be used in the Umbraco repositories here: #14564

**Comment by kjac** [comment]
Hi @nikcio 👋 

That's a mouthful 😄 thank you for voicing your concerns and taking the time to do so in a detailed manner. Let me try to (hopefully) address the main points here.

## NPoco migrations

First and foremost: You cannot skip a major version when upgrading a site. Say you're upgrading from v10 to v12, you _must_ first upgrade from v10 to v11 and then again from v11 to v12. This applies only at database level, meaning you don't have to fix obsoletions and breaking changes in the project code specifically for v11 if you're targeting v12.

I am quite certain this was part of [the docs](https://docs.umbraco.com/umbraco-cms/fundamentals/setup/upgrading) at some point, but it must have gone missing 😢 this needs mending.

We need to avoid breaking changes, so we can't change the existing migrations from NPoco to EF Core. For the same reason, we can't make a clean cut just now. While it would be really nice to do so, we must remain backwards compatible for two major versions. We do not have concrete plans to deprecate NPoco migrations just yet, but with the v12 release last month we're looking at v14 at the earliest to be "NPoco migration free". 

However - it shouldn't really matter. Migrations are atomic by nature, so whatever powers the individual migration should not affect other migrations.

## DbContext 

The `UmbracoDbContext` is not supposed to be used outside of the Umbraco code. If anything it should probably have been marked `internal` rather than `public` - we'll have to investigate that. Bottom line though: We consider `UmbracoDbContext` (and any DB models eventually referred by it) for internal usage only, just as we consider the actual DB tables internal.

Again this has to do with backwards compatibility. We can't (and won't) be backwards compatible at DB level. Developers are not supposed to be using `UmbracoDbContext` in their own code, just as they're not supposed to be querying the tables directly. This is why we made it possible to bring in custom `DbContext` implementations for those wanting to use EF Core rather than NPoco for their custom table queries.

## DbContext services

You're absolutely right about the pooled DB context. This needs fixing. Fortunately we can do that in a non-breaking way 😄 

## Provider setup

Despite the immediate gain in readability, I'm not a big fan of hardcoding assembly names as you've [done in your PR](https://github.com/umbraco/Umbraco-CMS/pull/14563/files#diff-308d9980336c9ca74e995c637a2dd1b1a3d862e8403c631782e4586edd35d68bR95-R115). However, using `optionsBuilder.MigrationsAssembly` seems nice and also adds to the readability. This is again something that we need to look into, and likely also something we can fix in a non-breaking way.

**Comment by JasonElkin** [comment]
Hey @kjac, upgrading via every major will need to be communicated very loud and clear if it's going to be the case going forwards. An interesting side effect of this approach means there is no way to actually stick to LTS-only installs, which I think some were hoping for.

It also means that the upgrade path between LTS versions will always involve upgrading via an EOL version of Umbraco (and an EOL version of .NET if done late in the support cycle). This has the potential to cause a few unnecessaryheadaches.

**Comment by nikcio** [comment]
And there shouldn't be any reason to limit migration between every major because the all the missing migrations will run when upgrading. It's only needed to have a hard limit on upgrading between majors when you have to remove an old migration provider for example 😉

**Comment by kjac** [comment]
@nikcio turning a gaze back at your original comment... We certainly don't want to keep building NPoco migrations if we can avoid it. But the future is notoriously hard to foresee, so I'd be silly to say it'll never happen again. I do agree it makes a ton of sense to stop writing them, because it'll only prolong the time we need to support them.

Other than that... is it safe to say that the only thing requiring immediate action is to look into pooled DB contexts?

**Comment by nikcio** [comment]
@kjac I would also argue that the missing DbSets/models for OpenIddict and the incorrect project references are quite critical - other than the project reference described above theres also incorrect references in Umbraco.Cms and the Umbraco integration test project. See the PR for more info.

The DbSet and models are really important to avoid developer confusion and the migrations and modelssnapshot references what class is used.

**Comment by nikcio** [comment]
This is how I see the `InitialCreate` migration work: https://github.com/umbraco/Umbraco-CMS/pull/14563/commits/a42baebf1d039ee0e541970e2a24cc95e6aceafd

**Comment by nul800sebastiaan** [comment]
Just to clarify, I've talked to @bergmania our policy is to support upgrading from LTS to LTS without having to go through each and every STS in between.

So you can go from 10 straight to 13 and then from 13 straight to 17 (the next LTS). We don't support an upgrade path from 10 to 17 though (and to 21 etc), you will need to go through each LTS.

**Comment by bergmania** [comment]
Hi @nikcio.

Sorry for the late reply - You hit directly into my vacation period 🌴.

Thanks you for all the work and expressing your concerns. Kenn already have addressed the most 😄 

We don't like to expose the _OpenIddict_ tables directly on `UmbracoDbContext`, instead we prefer to use the standard _OpenIddict_ methods.

Regarding your concerns about readability of the migrations, I agree it should have been named better. I will add an empty InitialCreate and an explicit `AddOpenIddict`. We don't want to add all existing tables yet as that will lead to double-maintenance every time a change to the tables are required.

Regarding the `OnModelCreating` that prepend "umbraco" to the tables, I think this is okay, and we can use the future move to EF Core for all repositories to ensure every table is prefixed "umbraco" instead of only a subset.

**Comment by nikcio** [comment]
>Regarding your concerns about readability of the migrations, I agree it should have been named better. I will add an empty InitialCreate and an explicit AddOpenIddict. We don't want to add all existing tables yet as that will lead to double-maintenance every time a change to the tables are required.

@bergmania 
It's really sad that you didn't do this before releasing 12.1 because this means that the `__EFMigrationHistory` will now contain a migration named `20230622183638_InitialCreate` that references the OpenIddict migration so if you rename it you also have to either add the new name to the migration history or do some SQL update in the migrations to rename this one. Otherwise, I'm pretty sure it will cause problems in EFCore as it will register it as not having been run 😞 

And having an actual InitialCreate migration, even an empty one, would have made so much more sense when we have to look at the migration history. It should have been the first one.
